File: android/EGETrainerApp/core/src/main/kotlin/ru/egetrainer/core/services/HomeworkService.kt
package ru.egetrainer.core.services

import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import ru.egetrainer.core.AppJson
import ru.egetrainer.core.models.AttemptPayload
import ru.egetrainer.core.models.Homework
import ru.egetrainer.core.models.HomeworkArchiveItem
import ru.egetrainer.core.models.HomeworkArchivePage
import ru.egetrainer.core.models.HomeworkAttempt
import ru.egetrainer.core.models.HomeworkSummary
import ru.egetrainer.core.models.StartAttemptResult
import ru.egetrainer.core.models.SubmitAttemptResult

/** Домен ДЗ ученика (зеркало app/providers/homework.js, порт HomeworkService.swift). */
class HomeworkService(private val client: SupabaseClient) {

    /** Список ДЗ ученика (student_my_homeworks_summary). */
    suspend fun myHomeworksSummary(): HomeworkSummary =
        client.rpc("student_my_homeworks_summary", deserializer = HomeworkSummary.serializer())

    /** ДЗ по токену ссылки (get_homework_by_token). */
    suspend fun homework(byToken: String): Homework =
        client.rpcSingleRow(
            "get_homework_by_token",
            buildJsonObject { put("p_token", byToken) },
            Homework.serializer(),
        )

    /** Начать/возобновить попытку (start_homework_attempt). */
    suspend fun startAttempt(token: String, studentName: String): StartAttemptResult {
        // student_key в RPC нормализуется на бэке; имя — как в вебе (trim + collapse spaces)
        val name = studentName.trim().split(Regex("\\s+")).filter { it.isNotEmpty() }.joinToString(" ")
        return client.rpcSingleRow(
            "start_homework_attempt",
            buildJsonObject {
                put("p_token", token)
                put("p_student_name", name)
            },
            StartAttemptResult.serializer(),
        )
    }

    /** Сдать ДЗ (submit_homework_attempt_v2 — контракт submitHomeworkAttempt веба). */
    suspend fun submitAttempt(
        attemptId: String,
        payload: AttemptPayload,
        total: Int,
        correct: Int,
        durationMs: Int,
    ): SubmitAttemptResult {
        val payloadJson = AppJson.encodeToJsonElement(AttemptPayload.serializer(), payload)
        return client.rpcSingleRow(
            "submit_homework_attempt_v2",
            buildJsonObject {
                put("p_attempt_id", attemptId)
                put("p_payload", payloadJson)
                put("p_total", total)
                put("p_correct", correct)
                put("p_duration_ms", durationMs)
            },
            SubmitAttemptResult.serializer(),
        )
    }

    /** Последняя попытка текущего ученика по токену (get_homework_attempt_by_token). */
    suspend fun attempt(byToken: String): HomeworkAttempt? =
        try {
            client.rpcSingleRow(
                "get_homework_attempt_by_token",
                buildJsonObject { put("p_token", byToken) },
                HomeworkAttempt.serializer(),
            )
        } catch (_: SupabaseError.EmptyResponse) {
            null
        }

    /**
     * Архив ДЗ с пагинацией (student_my_homeworks_archive — как
     * getStudentMyHomeworksArchive веба: p_offset/p_limit, веб начинает с offset=10).
     */
    suspend fun archive(offset: Int, limit: Int = 50): List<HomeworkArchiveItem> {
        val element = client.rpcElement(
            "student_my_homeworks_archive",
            buildJsonObject {
                put("p_offset", offset)
                put("p_limit", limit)
            },
        )
        return HomeworkArchivePage.parse(element)
    }

    /** Отчёт по попытке для учителя (get_homework_attempt_for_teacher). */
    suspend fun attemptForTeacher(attemptId: String): HomeworkAttempt =
        client.rpcSingleRow(
            "get_homework_attempt_for_teacher",
            buildJsonObject { put("p_attempt_id", attemptId) },
            HomeworkAttempt.serializer(),
        )
}

File: ios/EGETrainerApp/EGETrainerApp/Services/HomeworkService.swift
import Foundation

/// Домен ДЗ ученика (зеркало app/providers/homework.js).
struct HomeworkService {
    let client: SupabaseClient

    init(client: SupabaseClient = .shared) {
        self.client = client
    }

    /// Список ДЗ ученика (student_my_homeworks_summary).
    func myHomeworksSummary() async throws -> HomeworkSummary {
        try await client.rpc("student_my_homeworks_summary", params: [:], as: HomeworkSummary.self)
    }

    /// ДЗ по токену ссылки (get_homework_by_token).
    func homework(byToken token: String) async throws -> Homework {
        try await client.rpcSingleRow(
            "get_homework_by_token",
            params: ["p_token": .string(token)],
            as: Homework.self
        )
    }

    /// Начать/возобновить попытку (start_homework_attempt).
    func startAttempt(token: String, studentName: String) async throws -> StartAttemptResult {
        // student_key в RPC нормализуется на бэке; имя — как в вебе (trim + collapse spaces)
        let name = studentName
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .components(separatedBy: .whitespaces)
            .filter { !$0.isEmpty }
            .joined(separator: " ")
        return try await client.rpcSingleRow(
            "start_homework_attempt",
            params: ["p_token": .string(token), "p_student_name": .string(name)],
            as: StartAttemptResult.self
        )
    }

    /// Сдать ДЗ (submit_homework_attempt_v2 — контракт submitHomeworkAttempt веба).
    func submitAttempt(
        attemptId: String,
        payload: AttemptPayload,
        total: Int,
        correct: Int,
        durationMs: Int
    ) async throws -> SubmitAttemptResult {
        let payloadData = try JSONEncoder().encode(payload)
        let payloadJson = try JSONDecoder().decode(JSONValue.self, from: payloadData)
        let result = try await client.rpcSingleRow(
            "submit_homework_attempt_v2",
            params: [
                "p_attempt_id": .string(attemptId),
                "p_payload": payloadJson,
                "p_total": .number(Double(total)),
                "p_correct": .number(Double(correct)),
                "p_duration_ms": .number(Double(durationMs)),
            ],
            as: SubmitAttemptResult.self
        )
        // WIOS.2: состояние ученика изменилось → витрина локального подбора устарела.
        await PickSnapshotCache.shared.invalidateAll()
        await AccordionScreenCache.shared.invalidateAll()
        return result
    }

    /// Последняя попытка текущего ученика по токену (get_homework_attempt_by_token).
    func attempt(byToken token: String) async throws -> HomeworkAttempt? {
        do {
            return try await client.rpcSingleRow(
                "get_homework_attempt_by_token",
                params: ["p_token": .string(token)],
                as: HomeworkAttempt.self
            )
        } catch SupabaseError.emptyResponse {
            return nil
        }
    }

    /// Архив ДЗ с пагинацией (student_my_homeworks_archive — как
    /// getStudentMyHomeworksArchive веба: p_offset/p_limit, веб начинает с offset=10).
    func archive(offset: Int, limit: Int = 50) async throws -> [HomeworkArchiveItem] {
        let page = try await client.rpc(
            "student_my_homeworks_archive",
            params: [
                "p_offset": .number(Double(offset)),
                "p_limit": .number(Double(limit)),
            ],
            as: HomeworkArchivePage.self
        )
        return page.items
    }

    /// Отчёт по попытке для учителя (get_homework_attempt_for_teacher).
    func attemptForTeacher(attemptId: String) async throws -> HomeworkAttempt {
        try await client.rpcSingleRow(
            "get_homework_attempt_for_teacher",
            params: ["p_attempt_id": .string(attemptId)],
            as: HomeworkAttempt.self
        )
    }
}
